src/attribute_context.cc

```diff
@@ -0,0 +1,97 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "src/attribute_context.h"
+#include "google/protobuf/timestamp.pb.h"
+
+using ::google::protobuf::Map;
+using ::google::protobuf::Timestamp;
+
+namespace istio {
+namespace mixer_client {
+namespace {
+
+// TODO: add code to build context to reduce attributes.
+// Only check these attributes to build context.
+std::set<std::string> kContextSet = {"serviceName", "peerId", "location",
+                                     "apiName", "apiVersion"};
+
+// Convert timestamp from time_point to Timestamp
+Timestamp CreateTimestamp(std::chrono::system_clock::time_point tp) {
+  Timestamp time_stamp;
+  long long nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                        tp.time_since_epoch())
+                        .count();
+
+  time_stamp.set_seconds(nanos / 1000000000);
+  time_stamp.set_nanos(nanos % 1000000000);
+  return time_stamp;
+}
+
+}  // namespace
+
+bool AttributeContext::UpdateDictionary(const Attributes& attributes) {
+  int next_index = name_map_.size();
+  bool changed = false;
+  for (const auto& it : attributes.attributes) {
+    const std::string& name = it.first;
+    if (name_map_.find(name) == name_map_.end()) {
+      changed = true;
+      name_map_[name] = ++next_index;
+    }
+  }
+  return changed;
+}
+
+void AttributeContext::FillProto(const Attributes& attributes,
+                                 ::istio::mixer::v1::Attributes* pb) {
+  if (UpdateDictionary(attributes)) {
```

Comment:
Calling UpdateDictionary at the top means that we have to walk the set of attributes twice and lookup every attribute twice.

How about adding a check in the primary loop below that says "if the attribute is not in the dictionary, then update the dictionary". This would avoid the double looping and the double lookup.